### PR TITLE
(feat) Adding useDebounce to the react utils

### PR DESCRIPTION
**Pull request by vasharma05**
## Requirements
- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-dev.docs.openmrs.org/#/getting_started/contributing?id=your-pr-title-should-indicate-the-type-of-change-it-is) label. See existing PR titles for inspiration.

#### For changes to apps
- [ ]  My work conforms to the [**OpenMRS 3.0 Styleguide**](https://om.rs/styleguide) and **design documentation**.

#### If applicable
- [ ] My work includes tests or is validated by existing tests.
- [ ] I have updated the [esm-framework mock](https://github.com/openmrs/openmrs-esm-core/blob/main/packages/framework/esm-framework/mock.tsx) to reflect any API changes I have made.

## Summary

We use SWR everywhere to fetch information and control the number of requests, especially for searching. We also use lodash in many places for this, which is fine. However, one downside is that the search component is no longer a controlled input. This is because the state set by the search input will be set after a delay, and if that same state is passed to the search box, it will not work correctly in the UI. In some cases, having an uncontrolled input is acceptable, but in cases where there is an external "Clear results" button that clears the search input, it becomes tedious to handle.

Instead of all these workarounds for a single use case, why not have a hook that handles debouncing for us without requiring an uncontrolled input?

For a similar example, you can refer: https://github.com/openmrs/openmrs-esm-patient-chart/pull/1322

## Example
```tsx
import { useDebounceValue } from "@openmrs/esm-react-utils";

function MyComponent() {
  const [searchTerm, setSearchTerm] = useState('');
  const debouncedSearchTerm = useDebounce(searchTerm);
  const swrResult = useSWR(`/api/search?q=${debouncedSearchTerm}`) 

  return (
    <Search
      labelText={t('search', 'Search')}
      onChange={(e) => setSearchTerm(e.target.value)}
      value={searchTerm}
    />
  )
}
```

### CODE REVIEW

File: packages/framework/esm-react-utils/src/useDebounce.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
+import { useEffect, useState } from "react";
+
+/**
+ * This hook will help us out in getting the debounced value of a state variable,
+ * which then makes a request to the backend.
```
Comment (brandones) [inline packages/framework/esm-react-utils/src/useDebounce.ts:5]:
```suggestion
 * This hook debounces a state variable. That state variable can then be used as the
 * value of a controlled input, while the return value of this hook is used for making
 * a request.
```

File: packages/framework/esm-react-utils/src/useDebounce.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ * 
+ * For example,
+ *
```
Comment (brandones) [inline packages/framework/esm-react-utils/src/useDebounce.ts:8]:
```suggestion
 * @example
```

File: packages/framework/esm-react-utils/src/useDebounce.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
+import { useEffect, useState } from "react";
+
+/**
+ * This hook will help us out in getting the debounced value of a state variable,
+ * which then makes a request to the backend.
+ * 
+ * For example,
+ * 
+ * ```tsx
+ * import { useDebounceValue } from "@openmrs/esm-react-utils";
```
Comment (brandones) [inline packages/framework/esm-react-utils/src/useDebounce.ts:10]:
```suggestion
 * import { useDebounce } from "@openmrs/esm-react-utils";
```

File: packages/framework/esm-react-utils/src/useDebounce.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+ *   const debouncedSearchTerm = useDebounce(searchTerm);
+ *   const swrResult = useSWR(`/api/search?q=${debouncedSearchTerm}`) 
+ * 
+ *  return (
+ *    <Search
+        labelText={t('search', 'Search')}
+        onChange={(e) => setSearchTerm(e.target.value)}
+        value={searchTerm}
+      />
+ *  )
```
Comment (brandones) [inline packages/framework/esm-react-utils/src/useDebounce.ts:23]:
```suggestion
 *    <Search
 *      labelText={t('search', 'Search')}
 *      onChange={(e) => setSearchTerm(e.target.value)}
 *      value={searchTerm}
 *    />
 *  )
```

File: packages/framework/esm-react-utils/src/useDebounce.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+ * }
+ * 
+ * ```
+ * 
+ * @param value
+ * @param delay number = 300
+ * @returns debounceValue
+ */
```
Comment (brandones) [inline packages/framework/esm-react-utils/src/useDebounce.ts:31]:
```suggestion
 * ```
 * 
 * @param value: The value that will be used to set `debounceValue`
 * @param delay number = 300: The number of milliseconds to wait before updating `debounceValue`
 * @returns debounceValue: The debounced value
 */
```

**Comment by brandones** [comment]
You'll need to add the export to `public.ts` for it to be importable via `@openmrs/esm-framework`.

Please add this to the esm-framework mocks as well. You can just do `export { useDebounce } from '....wherever';`.

Please don't just check all the boxes in the PR. Only check them if they're both relevant and done. Styleguide: irrelevant. Tests: not done but optional for a hook this simple. Mock: required, not done.

**Comment by brandones** [comment]
Applied suggestions. You still need to add to public.ts.

**Comment by vasharma05** [comment]
I thought I did. Pushing it now

**Comment by vasharma05** [comment]
I did the changes but just had it locally only 🤦 

Please re-review
Thanks!

File: packages/framework/esm-framework/docs/API.md [hunk 6]
```diff
@@ -154,6 +154,7 @@
 ### Other Functions
 
 - [ExtensionSlot](API.md#extensionslot)
+- [useDebounce](API.md#usedebounce)
```
Comment (ibacher) [inline packages/framework/esm-framework/docs/API.md:157]: Could we categorise this as an API function?

File: packages/framework/esm-framework/docs/API.md [hunk 7]
```diff
@@ -3751,6 +3752,60 @@ Passing a function as children
...
+| Name | Type | Default value | Description |
+| :------ | :------ | :------ | :------ |
+| `value` | `T` | `undefined` | - |
+| `delay` | `number` | `300` | number = 300: The number of milliseconds to wait before updating `debounceValue` |
+
+#### Returns
+
+`T`
+
+debounceValue: The debounced value
```
Comment (ibacher) [inline packages/framework/esm-framework/docs/API.md:3801]: It looks like this comment isn't formatted correctly (note the "Default value" of `undefined` and the lack of description for `value` above. Also, I'm not sure the return docs are right.

**Review by brandones — APPROVED** [review]

**Pull request merged**